Produlytics/ui/back-end/src/main/java/it/deltax/produlytics/uibackend/common/detection/adapters/DetectionsPersistenceAdapter.java

```diff
@@ -0,0 +1,51 @@
+package it.deltax.produlytics.uibackend.common.detection.adapters;
+
+import it.deltax.produlytics.persistence.DetectionEntity;
+import it.deltax.produlytics.uibackend.common.detection.business.domain.DetectionLight;
+import it.deltax.produlytics.uibackend.common.detection.business.ports.out.ListDetectionsByCharacteristicPort;
+import it.deltax.produlytics.uibackend.repositories.RilevazioneRepository;
+import org.springframework.data.domain.Sort;
+import org.springframework.stereotype.Component;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@Component
+public class DetectionsPersistenceAdapter implements ListDetectionsByCharacteristicPort {
+
+    private final RilevazioneRepository repo;
+
+    public DetectionsPersistenceAdapter(RilevazioneRepository repo) {
+        this.repo = repo;
+    }
+
+    @Override
+    public List<DetectionLight> listByCharacteristic(int machineId, int characteristicId, Optional<Long> createdAfter) {
+        List<DetectionEntity> detections;
+        if (createdAfter.isPresent()) {
+            detections = repo.findByIdDeviceIdAndIdCharacteristicIdAndIdCreationTimeGreaterThan(
+                machineId,
+                characteristicId,
+                Instant.ofEpochMilli(createdAfter.get()),
+                Sort.by("creazioneUtc")
+            );
+        } else {
+            detections = repo.findByIdDeviceIdAndIdCharacteristicId(
+                machineId,
+                characteristicId,
+                Sort.by("creazioneUtc")
+            );
+        }
+        return detections.stream()
+            .map(rilevazione ->
+                new DetectionLight(
+                    rilevazione.getValue(),
+                    rilevazione.getId().getCreationTime().getEpochSecond(),
+                    rilevazione.getOutlier()
+                )
+            )
+            .collect(Collectors.toList());
+    }
```

Comment:
Ripensandoci questa non sarebbe logica di business?

Reply:
Snì 

Cioè, spring jpa non accetta valori opzionali per le query (oppure non ho scavato abbastanza) e mi pareva brutto fare due Ports per una cosa del genere

La strada giusta sarebbe cercare di costruire un metodo unico nel Repository con i valori opzionali @albertolazari @leidard  (magari spring boot lo fa già se gli passi null??)

Reply:
Secondo me invece è meglio fare due ports e metti la logica di quando usare una e quando l'altra nella parte di business. Conta poi che se vogliamo aggiornare le ultime rilevazioni quando otteniamo quelle nuove allora c'è ancora più logica e differenza tra le due porte.

Reply:
Vabbuon

Reply:
Alla fine ho risolto diversamente, mettendo il check in jpa ---> meno complessità in fase di progettazione

Reply:
In che senso metti il check in jpa?